File: src/feature-flags.ts
import { existsSync, readFileSync } from "fs";
import { resolve } from "path";
import type { MiddlewareHandler } from "hono";

const FEATURE_FLAGS_PATH = resolve(import.meta.dir, "..", "docs", "feature-flags.json");
const SERVICE_PROFILES_PATH = resolve(import.meta.dir, "..", "docs", "service-profiles.json");

export type FeatureFlagId =
  | "corporate-memory"
  | "broad-admin-assistant"
  | "optional-dashboards"
  | "office-miro-mcp"
  | "direct-browser-mcp"
  | "testbench";

export type FeatureSurface = {
  routes: string[];
  ui: string[];
  agents: string[];
  mcp_packs: string[];
};

export type FeatureFlagDefinition = {
  id: FeatureFlagId;
  description: string;
  default_enabled: boolean;
  surfaces: FeatureSurface;
};

export type FeatureFlagState = FeatureFlagDefinition & {
  enabled: boolean;
  enabled_by?: string;
  reason?: string;
};

export type FeatureFlagsResponse = {
  profile: string;
  source: string;
  features: FeatureFlagState[];
};

type FeatureCatalog = {
  schema_version: number;
  default_profile: string;
  flags: Record<string, Omit<FeatureFlagDefinition, "id">>;
  profiles: Record<string, { enabled_features?: string[] }>;
};

type ServiceProfileCatalog = {
  default_profile?: string;
  profiles?: Record<string, { enabled_features?: string[] }>;
};

type FeatureOverride = {
  enabled?: boolean;
  enabled_by?: string;
  reason?: string;
};

type FeatureOverrideFile = {
  enabled_features?: string[];
  features?: Record<string, FeatureOverride>;
};

const MCP_PACK_FEATURES: Record<string, FeatureFlagId> = {
  yonote: "corporate-memory",
  "yonote-read": "corporate-memory",
  memory: "corporate-memory",
  excel: "office-miro-mcp",
  word: "office-miro-mcp",
  "google-docs": "office-miro-mcp",
  "google-sheets": "office-miro-mcp",
  miro: "office-miro-mcp",
  "miro-api": "office-miro-mcp",
  puppeteer: "direct-browser-mcp",
};

function readJson<T>(path: string): T {
  return JSON.parse(readFileSync(path, "utf-8")) as T;
}

function parseCsv(value: string | undefined): string[] {
  if (!value) return [];
  return value.split(",").map(item => item.trim()).filter(Boolean);
}

function loadCatalog(path = FEATURE_FLAGS_PATH): FeatureCatalog {
  const catalog = readJson<FeatureCatalog>(path);
  if (catalog.schema_version !== 1) {
    throw new Error(`Unsupported feature flag schema_version=${catalog.schema_version}`);
  }
  return catalog;
}

function serviceProfileFromEnv(env: NodeJS.ProcessEnv): string {
  const explicit = env.KONOHA_FEATURE_PROFILE || env.KONOHA_SERVICE_PROFILE;
  if (explicit?.trim()) return explicit.trim();
  if (existsSync(SERVICE_PROFILES_PATH)) {
    const serviceProfiles = readJson<ServiceProfileCatalog>(SERVICE_PROFILES_PATH);
    if (serviceProfiles.default_profile) return serviceProfiles.default_profile;
  }
  return "prod-core";
}

function enabledFeaturesForProfile(profile: string, catalog: FeatureCatalog): Set<string> {
  if (existsSync(SERVICE_PROFILES_PATH)) {
    const serviceProfiles = readJson<ServiceProfileCatalog>(SERVICE_PROFILES_PATH);
    const serviceProfileFeatures = serviceProfiles.profiles?.[profile]?.enabled_features;
    if (Array.isArray(serviceProfileFeatures)) return new Set(serviceProfileFeatures.map(String));
  }
  return new Set((catalog.profiles[profile]?.enabled_features ?? []).map(String));
}

function applyOverrideFile(states: Map<string, FeatureFlagState>, path: string): void {
  if (!existsSync(path)) return;
  const raw = readJson<FeatureOverrideFile>(path);
  for (const id of raw.enabled_features ?? []) {
    const state = states.get(String(id));
    if (!state) continue;
    state.enabled = true;
    state.enabled_by = `file:${path}`;
    state.reason = "enabled_features override";
  }
  for (const [id, override] of Object.entries(raw.features ?? {})) {
    const state = states.get(id);
    if (!state) continue;
    state.enabled = override.enabled === true;
    state.enabled_by = override.enabled_by || `file:${path}`;
    state.reason = override.reason || "feature override file";
  }
}

export function resolveFeatureFlags(env: NodeJS.ProcessEnv = process.env): FeatureFlagsResponse {
  const catalog = loadCatalog();
  const profile = serviceProfileFromEnv(env);
  const profileEnabled = enabledFeaturesForProfile(profile, catalog);
  const states = new Map<string, FeatureFlagState>();

  for (const [id, definition] of Object.entries(catalog.flags)) {
    const enabled = definition.default_enabled || profileEnabled.has(id);
    states.set(id, {
      id: id as FeatureFlagId,
      ...definition,
      enabled,
      ...(enabled ? {
        enabled_by: profileEnabled.has(id) ? `service-profile:${profile}` : "catalog-default",
        reason: profileEnabled.has(id) ? "enabled by selected service profile" : "enabled by catalog default",
      } : {}),
    });
  }

  applyOverrideFile(states, env.KONOHA_FEATURE_FLAGS_FILE || "/opt/shared/konoha-feature-flags.json");

  const envEnabled = parseCsv(env.KONOHA_ENABLED_FEATURES);
  for (const id of envEnabled) {
    const state = states.get(id);
    if (!state) continue;
    state.enabled = true;
    state.enabled_by = env.USER ? `env:${env.USER}` : "env:KONOHA_ENABLED_FEATURES";
    state.reason = env.KONOHA_FEATURE_ENABLE_REASON || "enabled via KONOHA_ENABLED_FEATURES";
  }

  const envDisabled = parseCsv(env.KONOHA_DISABLED_FEATURES);
  for (const id of envDisabled) {
    const state = states.get(id);
    if (!state) continue;
    state.enabled = false;
    state.enabled_by = undefined;
    state.reason = undefined;
  }

  return {
    profile,
    source: FEATURE_FLAGS_PATH,
    features: [...states.values()].sort((a, b) => a.id.localeCompare(b.id)),
  };
}

export function isFeatureEnabled(id: FeatureFlagId, env: NodeJS.ProcessEnv = process.env): boolean {
  return resolveFeatureFlags(env).features.some(feature => feature.id === id && feature.enabled);
}

export function featureForMcpPack(packName: string): FeatureFlagId | undefined {
  return MCP_PACK_FEATURES[packName];
}

export function requireFeature(id: FeatureFlagId): MiddlewareHandler {
  return async (c, next) => {
    const feature = resolveFeatureFlags().features.find(item => item.id === id);
    if (!feature?.enabled) {
      return c.json({
        error: "Feature disabled",
        feature: id,
        reason: feature?.description || "This experimental surface is disabled by the selected Konoha profile.",
      }, 404);
    }
    await next();
  };
}

File: scripts/feature_flags.py
"""Shared Konoha feature flag contract for healthcheck and profile tooling."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from service_profiles import resolve_service_profile_from_env


FEATURE_FLAGS_PATH = Path(__file__).resolve().parents[1] / "docs" / "feature-flags.json"
FEATURE_FLAGS_OVERRIDE_PATH = Path("/opt/shared/konoha-feature-flags.json")


@dataclass(frozen=True)
class FeatureFlagState:
    id: str
    enabled: bool
    description: str
    enabled_by: str = ""
    reason: str = ""


def load_feature_flags(path: Path = FEATURE_FLAGS_PATH) -> dict[str, Any]:
    with path.open(encoding="utf-8") as f:
        raw = json.load(f)
    if raw.get("schema_version") != 1:
        raise ValueError(f"Unsupported feature flag schema_version={raw.get('schema_version')}")
    flags = raw.get("flags")
    if not isinstance(flags, dict) or not flags:
        raise ValueError("feature flag catalog must define flags")
    return raw


def _csv(value: str | None) -> set[str]:
    if not value:
        return set()
    return {item.strip() for item in value.split(",") if item.strip()}


def _apply_override_file(states: dict[str, FeatureFlagState], path: Path) -> None:
    if not path.exists():
        return
    raw = json.loads(path.read_text(encoding="utf-8"))
    for flag_id in raw.get("enabled_features") or []:
        state = states.get(str(flag_id))
        if not state:
            continue
        states[state.id] = FeatureFlagState(
            id=state.id,
            enabled=True,
            description=state.description,
            enabled_by=f"file:{path}",
            reason="enabled_features override",
        )
    for flag_id, override in (raw.get("features") or {}).items():
        state = states.get(str(flag_id))
        if not state:
            continue
        states[state.id] = FeatureFlagState(
            id=state.id,
            enabled=override.get("enabled") is True,
            description=state.description,
            enabled_by=str(override.get("enabled_by") or f"file:{path}"),
            reason=str(override.get("reason") or "feature override file"),
        )


def resolve_feature_flags(environ: dict[str, str] | None = None, path: Path = FEATURE_FLAGS_PATH) -> tuple[str, list[FeatureFlagState]]:
    env = environ or os.environ
    catalog = load_feature_flags(path)
    profile_env = dict(env)
    if profile_env.get("KONOHA_FEATURE_PROFILE"):
        profile_env["KONOHA_SERVICE_PROFILE"] = str(profile_env["KONOHA_FEATURE_PROFILE"])
    profile = resolve_service_profile_from_env(profile_env)
    profile_features = set(profile.enabled_features)

    states: dict[str, FeatureFlagState] = {}
    for flag_id, definition in catalog["flags"].items():
        enabled = bool(definition.get("default_enabled")) or flag_id in profile_features
        states[flag_id] = FeatureFlagState(
            id=flag_id,
            enabled=enabled,
            description=str(definition.get("description") or ""),
            enabled_by=f"service-profile:{profile.id}" if enabled and flag_id in profile_features else ("catalog-default" if enabled else ""),
            reason="enabled by selected service profile" if enabled and flag_id in profile_features else ("enabled by catalog default" if enabled else ""),
        )

    override_path = Path(env.get("KONOHA_FEATURE_FLAGS_FILE") or FEATURE_FLAGS_OVERRIDE_PATH)
    _apply_override_file(states, override_path)

    for flag_id in _csv(env.get("KONOHA_ENABLED_FEATURES")):
        state = states.get(flag_id)
        if not state:
            continue
        states[flag_id] = FeatureFlagState(
            id=state.id,
            enabled=True,
            description=state.description,
            enabled_by=f"env:{env.get('USER')}" if env.get("USER") else "env:KONOHA_ENABLED_FEATURES",
            reason=env.get("KONOHA_FEATURE_ENABLE_REASON") or "enabled via KONOHA_ENABLED_FEATURES",
        )

    for flag_id in _csv(env.get("KONOHA_DISABLED_FEATURES")):
        state = states.get(flag_id)
        if not state:
            continue
        states[flag_id] = FeatureFlagState(id=state.id, enabled=False, description=state.description)

    return profile.id, sorted(states.values(), key=lambda item: item.id)
